### ISSUE TITLE
App should completely restart after Google Play Services Update

### ISSUE THREAD

**Issue by uschindler** [corona-warn-app/cwa-app-android#513]
I installed the app for testing purposes on a device which was switched of for several months.

As expected, when app started at the screen for enabling the tracking, the message to update Google Play Services was displayed to the user. The link redirected to the Play Store where you were able to update "Google Play-Dienste" (Google Play Services) -> great!

After the upgrade succeeded, I switched back to CWA. Unfortunately, the app then says that something is not OK and the stack trace shown on "details" implies that Google Play Services 17.0.x is still in use (the version I had installed on the phone). I can't post a screenshot, because #77 was not fixed before release.

It looks like this is caused by the fact that the app wasn't restarted after the update, so it was still trying to link to the older version of the services apk. Enforce close of the app or restarting smartphone helps, but this is no good solution, the app should fully exit after the update of play services was triggered.

People on Twitter and Play Store comments are also complaining about that.

**Comment by HrGaertner** [comment]
I have the same problem but i was not fixed after a restart/reinstallation of the CWA. Further when i start the app it still says "you need to update your google play services" but my google play services are up to date

**Comment by uschindler** [comment]
Do you have enabled an Google Account on the phone? If you run the phone without link to Google, any updates from play store / services are disabled.

**Comment by HrGaertner** [comment]
Yes. It works now after multiple reboots/reinstallations

**Comment by mxmxchere** [comment]
> Do you have enabled an Google Account on the phone? If you run the phone without link to Google, any updates from play store / services are disabled.

Do you have any reference for me where this is written down? I dont have a Google-Account on my phone and the app isn't working for me.

**Comment by jakobmoellersap** [comment]
Hi @uschindler , thanks for your detailed input in this topic and many other issues! Unfortunately, we are not able to read out information on when GMS (and thus the necessary component for Exposure Notifications in Google Play Services) update. From what we know now, the updates are queued at random and cannot be influenced and/or observed from an external application. Sorry for the inconvenience.

**Comment by robertbastian** [comment]
Well, just because you can't restart the app programatically doesn't mean this isn't an issue. A solution would be to tell the user to restart the app once GMS is updated, and provide a restart button.

**Comment by uschindler** [comment]
Hi,

> Hi @uschindler , thanks for your detailed input in this topic and many other issues! Unfortunately, we are not able to read out information on when GMS (and thus the necessary component for Exposure Notifications in Google Play Services) update. From what we know now, the updates are queued at random and cannot be influenced and/or observed from an external application. Sorry for the inconvenience.

This is not completely right. According to the Google Play Store the update was successfully installed and ready. After that I moved back to the CWA app, so the new playstore service was already running, but obviously not linked to the app.

My suggestion (as @robertbastian also said). Once you triggered the messagebox that play services are outdated and the user clicks on the button to launch play store to update them (is this messagebox coming from you?) the CWA app should just stop and process should end.

If the user comes back, it would link and start completely new. If the user is doing this too early, it's not your fault.

**Comment by corneliusroemer** [comment]
Dear maintainers (@jakobmoellersap @tkowark @SebastianWolf-SAP),

It appears as though you are closing an issue despite the root cause not being fixed. As people have mentioned, while the automatic restart may not work, a button or better error information/troubleshooting info would go some way towards helping people not run into this issue and get stuck.

Please reopen or open another issue that addresses the same root issue with different wording that is more applicable.

The community and @SebastianWolf-SAP had a discussion yesterday about not closing issues that are still impacting people. It would be great if this information is spread widely across maintainers.

For me personally, it doesn't leave a good feeling behind when valid concerns are closed by maintainers against the community consensus. It feels heavy handed. The community should have a voice in deciding which issues matter.

If there is no consensus for closing, maybe you could suggest closing in the future before you close, so that you can listen to the community view. In a community, listening is very important.

Please acknowledge and respond to the points I raised. Thank you!

**Comment by uschindler** [comment]
Myaybe the actual installation is really delayed, although Google Play Store says it's uptodate.

To me it looked like a restart of phone helps in most cases. Maybe CWA should give the user this hint when the message box oocurs: "after updating google play services, wait some time and restart your phone to make sure the update is ready".

**Comment by uschindler** [comment]
Some other hints to solve the issue are:
- restart phone (as said before)
- clear cache of Google Play Services
- make sure you are logged in with your google account on the phone
This should be documented and maybe a link in the messagebox to some howto documentation should be added.

When going through Twitter yesterday evening, it looked to me that many people are really angry about that because the error messages are so cryptic and no information is given anywhere. I know it's not your fault, its often also user's own fault (like running a phone without a linked google account where play services updates are not installed).

**Comment by corneliusroemer** [comment]
So it's got relabelled but not reopened?

**Comment by kolyaopahle** [comment]
Hi,
sadly the message box mentioned here
> My suggestion (as @robertbastian also said). Once you triggered the messagebox that play services are outdated and the user clicks on the button to launch play store to update them (is this messagebox coming from you?) the CWA app should just stop and process should end.

is also not in our control.

Furthermore there is a difference between the GMS component of the Google Play Services updating and Google Play Services themselves updating. 
We could probalby detect the first case by waiting for a return intent (im not 100% sure about this, i have not tested it) but, as you already experienced, this does not guarantee that the GMS component was also updated.

**Comment by corneliusroemer** [comment]
Maintainers, your response is requested on the decision to close, see above @SebastianWolf-SAP @tkowark @jakobmoellersap

Since this is an issue, it's good to have it open and allow comments, maybe someone has a good idea how to deal with it - or escalates it to Google's Exposure API

**Comment by vaubaehn** [comment]
Hi @svengabr ,

digging in some old issues, I found this one. It's about some problems that occured after GMS was updated. The problems couldn't be solved in that days, because GSM was not providing any ENF version number. That feature has now been implemented in ENF 1.6, and could be used in general. #1150 will make use of it from CWA 1.4.0 on.
Maybe it's worth mirroring to Jira as an enhancement idea, to take control over version changes of ENF from inside of CWA?
However, since ENF version updates seem to run flawlessly for the time being, that would not be of high priority.

FYI: @thomasaugsten @jakobmoellersap
